### ISSUE TITLE
Contacts Sync with Duplicate Numbers

### ISSUE THREAD

**Issue by andyholmes** [andyholmes/gnome-shell-extension-gsconnect#441]
Hello,

I pressed the cross icon on Advanced menu (for contacts AND SMS). Then resync with the + button.
I search for the same contact (and i still have the two lines for phone number), and i have the conversation's history (is it normal ?).

My contact received the message.

_Originally posted by @gautierlabarre in https://github.com/andyholmes/gnome-shell-extension-gsconnect/issues/380#issuecomment-452746223_

**Comment by gautierlabarre** [comment]
> Can you confirm you don't have two phone numbers defined for these contacts? Once the contacts are cleared each number is only added once to the list. If there are two phone numbers with different formats but the same spacing, it's likely that's how GSConnect is receiving them in the vCard dump.

Yes, i checked more than one contact, they all have only one phone number saved on contacts.google.com (i also checked on my phone to be more thorough : one number also).

However, on the contact app, on my phone, at the bottom, there is a line saying "imported from" and two icons (with no tooltip) so i guess the first one is google contact and the second is whatsapp.
With that in mind, i checked the contacts with only one phone number on GSConnect to see if i have their whatapps and yes, i do.

**Comment by andyholmes** [comment]
Could you check the local cache in `~/.cache/gsconnect/<device-id>/contacts.json` to ensure these contacts only have one phone number? It shouldn't be the case that GSConnect will add phone numbers that aren't there or change the phone numbers before adding them.

I can't seem to reproduce this, so I'm not sure what could be causing it.

**Comment by gautierlabarre** [comment]
Interestingly i have pictures (from contacts, and from MMS) and a file _contacts.json_ **outside** the device-id folder (inside the gsconnect folder). I don't want to risk deleting it (unless you tell me i can).
So, inside the only folder (which is, i guess my device-id), i open the contacts.json and search for the contact you know. I have indeed two numbers : 

```
"3639i1903002d0df6770d.3789r38-41": {
    "id": "3639i1903002d0df6770d.3789r38-41",
    "name": "M",
    "numbers": [
      {
        "type": "cell",
        "value": "06 74 37 50 50"
      },
      {
        "type": "cell",
        "value": "0674375050"
      }
    ],
    "origin": "device",
    "timestamp": 949629426
  },
```

--- 

Out of curiosity, i opened the contacts.json **outside** the device-id folder. And look : 
```
"ef0d6d8929cd3b0538c2fff9fbae20fd4ce4f018": {
    "id": "ef0d6d8929cd3b0538c2fff9fbae20fd4ce4f018",
    "name": "M",
    "numbers": [
      {
        "value": "06 74 37 50 50",
        "type": "cell"
      }
    ],
    "origin": "folks"
  },
```
origin folks ? But it does have only one phone number ! They all have only one number. But the file seems to be outdated (a contact i added a few days ago is not listed).

--- 
I tried a few things on my own : 

- I checked my second gmail account to check if my contacts were not imported from both accounts.  But no, i don't have the same list of people and they don't have phone numbers. 
- I have an application named contacts https://wiki.gnome.org/Apps/Contacts on my Ubuntu, my contacts have one phone number each (the same as contacts.google.com). 

--- 
If you want, i can delete a contact from contacts.google.com, empty the cache from GSConnect and KDE connect (if you think it's useful), then click the + button, add the contact on contacts.google and then click the + button again.

**Comment by ferdnyc** [comment]
> Out of curiosity, i opened the contacts.json **outside** the device-id folder. And look :
>  [...]
> origin folks ?

> I have an application named contacts https://wiki.gnome.org/Apps/Contacts on my Ubuntu, my contacts have one phone number each (the same as contacts.google.com).

Folks is the Gnome library that manages contacts information. Gnome Contacts is an interface to that information. As the `gnome-contacts(1)` man page says:

> gnome-contacts uses the folks library to pull together contact information from various sources, including telepathy, evolution-data-server, your configured GNOME online accounts, etc.

The `$HOME/.cache/gsconnect/contacts.json` file is where gsconnect caches the data it pulls from Folks on the local system, whereas `$HOME/.cache/gsconnect/[device-id]/contacts.json` is the cache of numbers synced from the device.

But based on this:
```
    "numbers": [
      {
        "type": "cell",
        "value": "06 74 37 50 50"
      },
      {
        "type": "cell",
        "value": "0674375050"
      }
    ],
    "origin": "device",
```
It _looks_ like your device is just sending through the same number twice, formatted differently.

I wonder if... one thing I've noticed on my own phone, recently, is that Google auto-added a "Duo Preview" account (listed in Settings > Accounts), even though I've never used Google Duo. It has the same credentials as my Google account, and Contacts are also being synced there, but it seems to have different ideas about some of the contacts' details than my Google account's Contacts list does. That's caused some issues (duplicate listings, mostly) in my Android Contacts app data, since it aggregates from all of the available Contacts sources.

Perhaps the two different formats for the same number coming off your device are due to something like that, a second contacts-management account that's storing the data with a different format?

**Comment by gautierlabarre** [comment]
Hello ferdnyc and thanks for the informations about gnome contacts.
I checked your idea, and i too have google Duo syncing (i stopped the auto sync) and yeah after i opened the app i see "the" contact (M) listed with two phone numbers (differently formatted again) : 

- +36 74 375 050 (weird format)
- 06 74 37 50 50

But it's not the case for every contacts on GSConnect that have more than one phone number, for instance : 

on GSConnect i have a contact saved as : 066-309-5050 and 0663095050. On duo it's just 06 63 09 50 50.

Well if every app have a different formatting system, i understand it's a true headache to solve :frowning_face:

**Comment by gautierlabarre** [comment]
For information, i deleted a contact and recreated it from contacts.google.com and it works, it has only one phone number now. 
Before i did so, i exported its informations in a .csv and after i created the new contact i compared the two csv and they are identical...

Maybe there is a conflict with another app like duo or whastapp like ferdnyc suggested, i'll keep you updated on this contact (if another line appears after something syncs)

**Comment by gautierlabarre** [comment]
The folder was deleted, inside the .cache folder there was no gsconnect folder. I tried to reproduce this, without success. So I might have done something else... Anyways, it did not alter the way the app was behaving and it happened under unusual manipulation (edit contact, delete cache, deactivate gsconnect extension from https://extensions.gnome.org/local/ and reactivate it.), so i won't create a bug for that...

On the other hand, the contact I deleted and added as new contact with only one phone number, now has 2 phones (06 75 05 50 50 and 0675055050). So, i guess it's something else syncing and altering the contact on my phone ?
This contact (on my phone) is associated with Gmail and Whatsapp. I did not add the whatsapp account myself, besides, I never talked to him on whatsapp.

But that's really weird, because it's only doing it for some contacts... Some of them have one phone number and are on whatsapp, duo and gmail... 

---

But wait, i think i found the logic behind this issue :

All the contacts with **one phone number have no spacing**, ex : 0614235050 or +33642225050. Some of them are on whatsapps, or google duo, but not all.
All the contacts with spaces and only ONE phone number are **not** on whatsapp. 
The only contact with associated account is on Google duo and Gmail. (so it seems duo is not rewriting phone number)
And all the contacts with two phone numbers are on whatsapp.

I checked ALL my contacts, the rule I set before is working on each of them.
